[Código o Número de la Tarjeta] Implementar Operaciones Con Lenguajes

gabrielhuav opened the pull request:
# [Código o Número de la Tarjeta] Implementar Práctica 2: Lenguajes y Autómatas con Software Interactivo

Este repositorio contiene el código para la **Práctica 2** de la asignatura **Teoría de la Computación**, donde se implementan operaciones sobre lenguajes formales y la simulación de autómatas finitos deterministas (AFD) utilizando la biblioteca **Flet** en Python.

## Instrucciones mínimas para poner en ejecución el proyecto

### Requisitos previos:

- **Python 3.7+**: Asegúrese de tener Python 3.7 o superior instalado.
- **Flet**: Biblioteca de Python para crear la interfaz gráfica.
- **Visual Studio Code** o cualquier otro editor de su preferencia.

### Instalación:

1. **Clonar este repositorio**:
   ```bash
   git clone https://github.com/usuario/repo-practica2.git
   cd repo-practica2
   ```
2. **Crear y activar un entorno virtual:**:
En Windows:

   ```bash
   git clone https://github.com/usuario/repo-practica2.git
   cd repo-practica2
   ```
    En macOS/Linux:
    
     ```bash
    python3 -m venv venv
    source venv/bin/activate
     ```
   
3. **Instalar las dependencias necesarias:**:
     ```bash
    pip install flet
     ```
4. **Ejecutar el programa::**:
     ```bash
    python app.py
     ```

## Uso del programa:

1. Definir el alfabeto: Ingrese el alfabeto separado por comas (ej: a,b,c).
2. Operaciones sobre lenguajes: Seleccione una operación, como Cerradura de Kleene, Concatenación, Unión, etc.
3. Definir AFD: Puede definir un autómata finito determinista ingresando los estados, transiciones y estados de aceptación.
4. Validación de cadenas: Valide si una cadena es aceptada o rechazada por el autómata.

## Capturas de pantalla:

1. Definir el alfabeto: Ingrese el alfabeto separado por comas (ej: a,b,c).
2. Operaciones sobre lenguajes: Seleccione una operación, como Cerradura de Kleene, Concatenación, Unión, etc.
3. Definir AFD: Puede definir un autómata finito determinista ingresando los estados, transiciones y estados de aceptación.
4. Validación de cadenas: Valide si una cadena es aceptada o rechazada por el autómata.

![image](https://github.com/user-attachments/assets/3359dbc6-9562-48cc-95ae-ef02b6195b90)
![image](https://github.com/user-attachments/assets/20016475-a0b7-425f-91d3-b4f7ff9e3a37)
![image](https://github.com/user-attachments/assets/efc1ef1c-2bea-40f9-99db-30274c52aa20)
![image](https://github.com/user-attachments/assets/f6ce7932-b5af-4759-a75b-16497c5dd854)

Proceso de QA:

- [x] Pruebas de QA realizadas.
- [x] Pruebas de QA aprovadas.

## Contribuciones:

Las contribuciones son bienvenidas. Por favor, abra un issue antes de enviar un pull request.

gabrielhuav committed:
Implementar Operaciones Con Lenguajes

chrisalmonte commented:
Falta Implementar la operación de unión.

chrisalmonte commented on practica2.py:
```diff
@@ -99,146 +118,61 @@ def main():
-            n = int(input("Ingrese el número máximo de combinaciones para la cerradura de Kleene: "))
-            kleene = kleene_cerradura(alfabeto, n)
-            print(f"Cerradura de Kleene hasta longitud {n}: {kleene}")
+        elif opcion == "7":
+            lenguaje1 = input("Ingrese las cadenas del primer lenguaje separadas por comas: ").split(',')
+            lenguaje2 = input("Ingrese las cadenas del segundo lenguaje separadas por comas: ").split(',')
+            resultado = interseccion_lenguajes(lenguaje1, lenguaje2)
+            print(f"Resultado de la intersección: {resultado}")
```
```suggestion
        elif opcion == "7":
            lenguaje1 = input("Ingrese las cadenas del primer lenguaje separadas por comas: ").split(',')
            lenguaje2 = input("Ingrese las cadenas del segundo lenguaje separadas por comas: ").split(',')
            resultado = interseccion_lenguajes(lenguaje1, lenguaje2)
            print("Resultado de la intersección: {resultado}")
```
f es inecesario

gabrielhuav committed:
Implementar AFD

Implementar AFD mediante su definición

chrisalmonte commented:
* No es posible ver elementos cuando sobrepasan el tamaño de la ventana.
![imagen](https://github.com/user-attachments/assets/6cf916e7-ea27-4242-aa09-b623ab4f7c94)

* Evaluar la definición de un autómata sin llenar los campos resulta en una aprobación. (¿Es comportamiento esperado?)
![imagen](https://github.com/user-attachments/assets/f7f9fa4e-96ba-4cf1-a794-c6afb37b0d24)

gabrielhuav committed:
Implementar comportamiento responsivo

Arreglar elementos fuera del tamaño de la ventana y comportamiento para definir el autómata completo antes de validar

gabrielhuav commented:
Ahora es posible ver elementos cuando sobrepasan el tamaño de la ventana:

![image](https://github.com/user-attachments/assets/281dab35-14df-4062-ac27-98025d6d2a2e)
![image](https://github.com/user-attachments/assets/cece921b-fcdc-4df1-8e3f-d4ed0dc1bcea)

Ahora hay que definir el autómata completo antes de validar.

![image](https://github.com/user-attachments/assets/94d79628-4121-4e86-92fd-2e071e1a38de)

chrisalmonte commented:
QA Completo.
